hack/generate-manifest-collector.sh

```diff
@@ -0,0 +1,140 @@
+running ipfix collector. You can set the HELM environment variable to the path of the
+helm binary you want us to use. Otherwise we will look for helm in your PATH and your
+GOPATH. If we cannot find helm there, we will try to install it."
+
+while [[ $# -gt 0 ]]
+do
+key="$1"
+
+case $key in
+    --mode)
+    MODE="$2"
+    shift 2
+    ;;
+    --port)
+    PORT="$2"
+    shift 2
+    ;;
+    --proto)
+    PROTO="$2"
+    shift 2
+    ;;
+    -h|--help)
+    print_usage
+    exit 0
+    ;;
+    *)    # unknown option
+    echoerr "Unknown option $1"
+    exit 1
+    ;;
+esac
+done
+
+if [ "$MODE" != "dev" ] && [ "$MODE" != "release" ]; then
+    echoerr "--mode must be one of 'dev' or 'release'"
+    print_help
+    exit 1
+fi
+
+if [ "$MODE" == "release" ] && [ -z "$IMG_TAG" ]; then
+    echoerr "In 'release' mode, environment variable IMG_TAG must be set"
+    print_help
+    exit 1
+fi
+
+THIS_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+
+source $THIS_DIR/verify-helm.sh
+
+if [ -z "$HELM" ]; then
+    HELM="$(verify_helm)"
+elif ! $HELM version > /dev/null 2>&1; then
+    echoerr "$HELM does not appear to be a valid helm binary"
+    print_help
+    exit 1
+fi
+
+cd $THIS_DIR/../build/yamls/collector
+
+# to support sed to run both on OSX, Linux and windows
+if [[ "$OSTYPE" == "darwin"* ]]; then
+    if [ "$MODE" == "dev" ]; then
+        sed -i '' -e "s/^image_tag.*/image_tag: latest/g" values.yaml
+        if [[ $PORT != "" ]]; then
+            sed -i '' -e "s/^port.*/port: $PORT/g" values.yaml
+        fi
+        if [[ $PROTO != "" ]]; then
+            sed -i '' -e "s/^protocol.*/protocol: $PROTO/g" values.yaml
+        fi
+        $HELM template "." -f "./templates/ipfix-collector.yaml"
+    fi
+
+    if [ "$MODE" == "release" ]; then
+        # update the Chart version by the newest release version
+        sed -i '' -e "s/^version.*/version: ${IMG_TAG:1}/g" Chart.yaml
+        # replace the line starting with "image_tag", and create a backup
+        sed -i.bak -e "s/^image_tag.*/image_tag: $IMG_TAG/g" values.yaml
+
+        $HELM template "." -f "./templates/ipfix-collector.yaml"
+        cp -f "values.yaml.bak" "values.yaml"
+        rm "values.yaml.bak"
+    fi
+else
+    if [ "$MODE" == "dev" ]; then
+        sed -i -e "s/^image_tag.*/image_tag: latest/g" values.yaml
+        if [[ $PORT != "" ]]; then
+            sed -i -e "s/^port.*/port: $PORT/g" values.yaml
+        fi
+        if [[ $PROTO != "" ]]; then
+            sed -i -e "s/^protocol.*/protocol: $PROTO/g" values.yaml
+        fi
+        $HELM template "." -f "./templates/ipfix-collector.yaml"
+    fi
+
+    if [ "$MODE" == "release" ]; then
+        sed -i -e "s/^version.*/version: ${IMG_TAG:1}/g" Chart.yaml
+        sed -i.bak -e "s/^image_tag.*/image_tag: $IMG_TAG/g" values.yaml
+
+        $HELM template "." -f "./templates/ipfix-collector.yaml"
+        cp -f "values.yaml.bak" "values.yaml"
+        rm "values.yaml.bak"
+    fi
+fi
```

Comment:
I can't figure out a good way to avoid the code repetition here. I tried a couple of ways but they didn't work.
For example introducing a variable:
`SEDOPTION=
if [[ "$OSTYPE" == "darwin"* ]]; then
  SEDOPTION="''"
fi

if [ "$MODE" == "dev" ]; then
    sed -i ${SEDOPTION} -e "s/^image_tag.*/image_tag: latest/g" values.yaml`
This would create an extra weird file named `values''.yaml`.

Reply:
I am not a sed expert.
How about setting SEDOPTION="-i" or "-i ''" depending on OSTYPE?

Reply:
Thanks Shih-Hao for the suggestion! I tried it but sadly the results is the same:( still bringing the extra file

Reply:
Just curious, is extra file caused by "sed -i.bak" on line 114, 134 or elsewhere?

Reply:
I'm testing it with command `/generate-manifest-collector.sh --mode dev > build/yamls/ipfix-collector.yaml`. 
So I think it goes to line 100, which is the code below with the variable `SEDOPTION="-i ''`:
`sed ${SEDOPTION} -e "s/^image_tag.*/image_tag: latest/g" values.yaml`

I think I can understand the logic that sed will create a file ending with '', just like what I have in line 116, where sed will create a backup file values.yaml.bak. But the strange thing is, without the SEDOPTION, the original code on line 100 works well on mac and will not create the extra file, but with SEDOPTION, it creates the file.

Reply:
Thanks for the info. 
One dumb question. So we need to pass "" after -i on windows machine?

Reply:
No, for MacOS we do `sed -i '' -e`, for Linux and Windows we do `sed -i -e`.

Reply:
Got it now. Thanks.
Maybe we can always use a dummy backup file for all the sed commands (e.g. "sed -i.dummy ..."), except "sed -i.bak ...".
Then in the end, remove the dummy file.

Reply:
Good idea! Thanks a lot. It works now.